Gems/RecastNavigation/Code/Source/Components/RecastNavigationMeshCommon.cpp

```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include "RecastNavigationMeshCommon.h"
+
+#include <DetourNavMeshBuilder.h>
+#include <AzCore/Console/Console.h>
+#include <AzCore/Debug/Profiler.h>
+#include <AzFramework/Entity/EntityDebugDisplayBus.h>
+#include <AzFramework/Physics/PhysicsScene.h>
+#include <RecastNavigation/RecastNavigationSurveyorBus.h>
+
+AZ_DEFINE_BUDGET(Navigation);
+
+namespace RecastNavigation
+{
+    NavigationTileData RecastNavigationMeshCommon::CreateNavigationTile(TileGeometry* geom,
+        const RecastNavigationMeshConfig& meshConfig, rcContext* context)
+    {
+        AZ_PROFILE_SCOPE(Navigation, "Navigation: create tile");
+
+        rcConfig config = {};
+        AZStd::vector<AZ::u8> trianglesAreas;
+        RecastPointer<rcHeightfield> solid;
+        RecastPointer<rcCompactHeightfield> compactHeightfield;
+        RecastPointer<rcContourSet> contourSet;
+        RecastPointer<rcPolyMesh> polyMesh;
+        RecastPointer<rcPolyMeshDetail> polyMeshDetail;
+
+        const float* vertices = &geom->m_vertices.front().m_x;
+        const int vertexCount = static_cast<int>(geom->m_vertices.size());
+        const int* triangleData = &geom->m_indices[0];
+        const int triangleCount = static_cast<int>(geom->m_indices.size()) / 3;
+
+        //
+        // Step 1. Initialize build config.
+        //
+
+        // Init build configuration from GUI
+        memset(&config, 0, sizeof(config));
+        config.cs = meshConfig.m_cellSize;
+        config.ch = meshConfig.m_cellHeight;
+        config.walkableSlopeAngle = meshConfig.m_agentMaxSlope;
+        config.walkableHeight = static_cast<int>(ceilf(meshConfig.m_agentHeight / config.ch));
+        config.walkableClimb = static_cast<int>(floorf(meshConfig.m_agentMaxClimb / config.ch));
+        config.walkableRadius = static_cast<int>(ceilf(meshConfig.m_agentRadius / config.cs));
+        config.maxEdgeLen = static_cast<int>(meshConfig.m_edgeMaxLen / meshConfig.m_cellSize);
```

Comment:
update: We probably want to use `aznumeric_cast` here instead and also prefer `AZStd::ceil`, `AZStd::floor` etc.. that use the correct c++ overloads

Reply:
done